File: Introductory Problems/Palindrome Reorder/code.js
const fs = require('fs');
const s = fs.readFileSync(0, 'utf-8').trim().split('\n')[0];

const count = {}
for (ch of s) {
    count[ch] = (count[ch] || 0) + 1;
}

const half  = []
let middle = ""
let odd_count = 0

const keys = Object.keys(count);

for (ch of keys) {
    const cnt = count[ch];
    if (cnt&1) {
        odd_count += 1;
        if (odd_count > 1) {
            console.log("NO SOLUTION");
            process.exit(0);
        }
        middle = ch;
    }
    half.push(ch.repeat(Math.floor(cnt/2)));
}

const half_str = half.join("");
const result = half_str + middle + [...half_str].reverse().join('');
console.log(result)

File: Introductory Problems/Palindrome Reorder/code.rs
use std::collections::HashMap;
use std::io::{self, BufRead};

fn main() {
    let stdin = io::stdin();
    let s = stdin.lock().lines().next().unwrap().unwrap();
    let s = s.trim();

    // Frequency Map
    let mut count:HashMap<char, usize> = HashMap::new();
    for ch in s.chars() {
        *count.entry(ch).or_insert(0) += 1;
    }

    let chars: Vec<char> = count.keys().copied().collect();
    let mut half = String::new();
    let mut middle = String::new();

    for ch in chars {
        let cnt = count[&ch];
        if cnt % 2 == 1 {
            if !middle.is_empty() {
                println!("NO SOLUTION");
                return;
            }
            middle.push(ch);
        }
        half.push_str(&ch.to_string().repeat(cnt / 2));
    }

    let reversed_half: String = half.chars().rev().collect();
    println!("{}{}{}", half, middle, reversed_half);

}